### DISCUSSION TITLE
How was solved the "conformance problem" in the AWS s3 implementation?

### DISCUSSION THREAD

**Discussion by gabrielgiussi** [p-org/P#577] [Q&A]
In the [twitch presentation](https://www.twitch.tv/videos/962963706?t=0h13m52s) Serdar Tarisan mentions the "conformance problem" which is to check the correspondence between the implementation and the design model. 

Ankush also talks about this as an unsolved problem in [this talk](https://youtu.be/5YjsSDDWFDY?t=1848), however Tarisan explains that they use the design "as an executable reference model and run it alongside the implementation code" in what he calls "code conformance check".

![image](https://user-images.githubusercontent.com/7468185/232055362-715702f3-d2c0-4800-8a30-5a6b814c5d8a.png)

Can you explain a bit more of this process? Hot the design becomes an executable reference model? How it runs "alongside" the implementation code?

**Comment by amw-zero** [comment]
Plus one on this, I came to ask the same question. Can P be used at all to check properties / spec conformance on the implementation level? Or is it primarily targeted at design verification?

**Reply by timmyjose** [reply]
I would imagine that it's impossible to check the conformance of the implementation against the model, in general. Verification tools like TLA+, P etc. can only explore the model itself. So also for Theorem Provers like Dafny (as shown in the first twitch talk slide) and Isabelle/HOL etc. Unless, of course, your whole software was written in Dafny or Isabelle/HOL, which is not feasible.

**Comment by gabrielgiussi** [comment]
I think [PObserve](https://youtu.be/pnfrWPFWbAA) might be the response.
